web/src/util.js

```diff
@@ -1,5 +1,254 @@
-const example = x => x
+import { useState } from "react"
 
-export {
-    example
-}
\ No newline at end of file
+const headerContentTypes = [
+    "application/json",
+    "application/xml",
+    "application/javascript",
+    "application/pdf",
+    "application/zip",
+    "application/octet-stream",
+    "application/ld+json",
+    "application/vnd.api+json",
+    "application/msword",
+    "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
+    "application/vnd.ms-excel",
+    "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
+    "application/vnd.ms-powerpoint",
+    "application/vnd.openxmlformats-officedocument.presentationml.presentation",
+    "application/x-www-form-urlencoded",
+    "application/x-tar",
+    "application/x-rar-compressed",
+    "application/x-7z-compressed",
+    "audio/mpeg",
+    "audio/wav",
+    "image/jpeg",
+    "image/png",
+    "image/gif",
+    "image/webp",
+    "image/svg+xml",
+    "multipart/form-data",
+    "text/html",
+    "text/plain",
+    "text/css",
+    "text/javascript",
+    "text/xml",
+    "text/csv",
+    "text/markdown",
+    "video/mp4",
+    "video/mpeg",
+    "video/webm",
+    "video/ogg"
+]
+
+const emptyValue = "<Enter Value>"
+
+const highlightPartOfString = (found, string) => {
+    const index = string.toLowerCase().indexOf(found.toLowerCase())
+    const [prefix, toHighlight, postfix] = [
+        string.slice(0, index),
+        string.slice(index, found.length + index),
+        string.slice(found.length + index, string.length)
+    ]
+    return { prefix, toHighlight, postfix }
+}
+/**
+ *
+ * @param {*} recommendations Options of Autocomplete as an array
+ * @returns
+ */
+const AutoCompleteInput = ({ recommendations }) => {
+    const [inputValue, setInputValue] = useState("")
+    const [firstRecommendation, setFirstRecommendation] = useState("")
+    const [showFirstRecommendation, setShowFirstRecommendation] = useState(true)
+    const [listActive, setListActive] = useState(false)
+    const [optionPointerIndex, setOptionPointerIndex] = useState(-1)
+    const [scrollWidth, setScrollWidth] = useState(null)
+    const [highlightedParts, setHighlightedParts] = useState([])
+    const [filteredRecommendations, setFilteredRecommendations] = useState(
+        recommendations.map((rec) => "p_" + rec)
+    )
+
+    const handleInputValueChange = (value) => {
+        if (value === undefined) return
+
+        if (!listActive) setListActive(true)
+        value = removePrefix(value)
+        setInputValue(value)
+        filterRecommendations(value)
```

Comment:
das brauchen wir nicht mehr, wenn wir useMemo nutzen. In React muss man etwas umdenken: da triggert man i.d.R. lieber ein neues Rendern der Komponente über einen state-setter und spart sich oft einige Logik, die innerhalb der Komponente irgendwas updated. Gibt natürlich Fälle, wo das dann doch zu teuer ist, aber da solltest du dir erstmal keinen Kopf drüber machen.